### PR TITLE
feat: Add WalletConnect QR code to ConnectWalletDialog with latest web3-react

### PR DESCRIPTION
**Pull request by kristiehuang**
Adds QR code functionality to WalletConnect. Also updates web3-react to latest version.

Known minor issues:
* After MetaMask disconnection, widget will still use MetaMask's JSON RPC connector to get price quotes. Doesn't cause issues in user experience tho. This is bc of web3-react's `useIsActive()` hook not working entirely as we'd expect it to
* Does not re-render when switching between accounts (also bc of web3-react's `useAccount()` hook)

Demo of WalletConnect QR working for both the popup modal & the tile QR:

qr

### CODE REVIEW

File: package.json
Comment (zzmp) [inline package.json:159]: These (@walletconenct/ethereum-provider, @web3-react/metamask, @web3-react/walletconnect) should be moved to dependencies (not devDeps).

File: src/cosmos/useJsonRpcEndpoint.ts
Comment (zzmp) [inline src/cosmos/useJsonRpcEndpoint.ts:10]: Rm `string` - just `string[]`.

File: src/cosmos/useJsonRpcEndpoint.ts
Comment (zzmp) [inline src/cosmos/useJsonRpcEndpoint.ts:38]: What does `at` compile to? If it compiles to `at`, and not `[]`, then prefer to use `[]` directly (as it is more widely available).

**Review by zzmp — CHANGES_REQUESTED** [review]
Should merging be deferred until https://github.com/NoahZinsmeister/web3-react/issues/591 is resolved, so that this doesn't bloat the widget size?

**Comment by zzmp** [comment]
Design question: Should the "Connect wallet to swap" button be more button-like? Should the text change at all?

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (zzmp) [inline src/hooks/connectWeb3/useActiveWeb3React.tsx:113]: I'm afraid this might cause frequent updates. Can the updater be moved to a separate context?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (zzmp) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:127]: Is there a safer way to do these html transforms?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (zzmp) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:124]: The QRCode that we end up with is not as pretty as the WC QR code. Do you know why?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (zzmp) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:68]: You switch between wcTileConnection and wcConnectionTile - stick with one name. They're in different contexts, so they can be re-used.

Comment (zzmp) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:68]: The important thing here is wc, not tile.

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (kristiehuang) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:127]: wdym by safer?

unfortunately this is literally how WalletConnect does it in their code - https://github.com/WalletConnect/walletconnect-legacy/blob/4563223a9fe2ff099e7a66788813697e09b1e35c/packages/modal/src/browser/components/QRCodeDisplay.tsx#L12

**Comment by kristiehuang** [comment]
> Should merging be deferred until [NoahZinsmeister/web3-react#591](https://github.com/NoahZinsmeister/web3-react/issues/591) is resolved, so that this doesn't bloat the widget size?

is there a plan for this to be resolved? I don't want to be blocked by an issue that's on the backlog

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (kristiehuang) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx:113]: hm - could you clarify this? wdym by "risk extra renders by instantiating objects multiple times"

**Comment by kristiehuang** [comment]
> Design question: Should the "Connect wallet to swap" button be more button-like? Should the text change at all?

pinged Phil about this!

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:124]: I think what you have currently complies with the designs: https://www.figma.com/file/A7w2snzxyFK2qj2aFpkkm1/Widgets?node-id=6057%3A39728

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (JFrankfurt) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:143]: `<Trans>{dynamicContent}</Trans>` doesn't work.

File: src/components/ConnectWallet/index.tsx
Comment (JFrankfurt) [inline src/components/ConnectWallet/index.tsx:1]: this is unused

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (JFrankfurt) [inline src/hooks/connectWeb3/useActiveWeb3React.tsx:141]: image

I think you should add this to the global window type

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (JFrankfurt) [inline src/hooks/connectWeb3/useActiveWeb3React.tsx:200]: @zzmp can you take a look at this?

@kristiehuang The name here makes it seem like this is a hook (`use______`), and it's using hooks internally, but the usage in `ConnectWalletDialog` is as a factory for connection handlers.

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (kristiehuang) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:71]: no :(

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (kristiehuang) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx:141]: I'm not getting this warning - are you using `"@types/node": "^16.7.13"` ?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (kristiehuang) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:174]: works fine - but depends on w3r's `usePriorityConnector` hook, which returns "the first of the initializedConnectors that is active"

i.e. if `const initializedConnectors = [metaMask, walletConnect]`
* I connect WalletConnect first, then also connect MetaMask -- it switches to MetaMask wallet
* I connect MetaMask first, then also connect WalletConnect -- no switch is made

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (kristiehuang) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx]: could I just rename it to `useWalletCallback`?

Comment (zzmp) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx]: Usage is correct, but naming is off (and I think you inherited this).
@kristiehuang could you change `useWallet`->`activateWallet`, and `useConnect`->`useConnectCallback`?

**Comment by zzmp** [comment]
This does not build for me locally. Note that the checks are not being run because this is not a PR into `main`.

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (zzmp) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx:141]: > > Should merging be deferred until [NoahZinsmeister/web3-react#591](https://github.com/NoahZinsmeister/web3-react/issues/591) is resolved, so that this doesn't bloat the widget size?
> 
> is there a plan for this to be resolved? I don't want to be blocked by an issue that's on the backlog

We've gotten pushback from integrators in the past about the size of this package, so I would investigate this before moving forward. It shouldn't block this PR, but it may need to block wallet-connect-flow from merging into main, or it may need to be noted in the release notes.

**Comment by zzmp** [comment]
> > Should merging be deferred until [NoahZinsmeister/web3-react#591](https://github.com/NoahZinsmeister/web3-react/issues/591) is resolved, so that this doesn't bloat the widget size?
> 
> is there a plan for this to be resolved? I don't want to be blocked by an issue that's on the backlog

We've gotten pushback from integrators in the past about the size of this package, so I would investigate this before moving forward. It shouldn't block this PR, but it may need to block wallet-connect-flow from merging into main, or it may need to be noted in the release notes.

File: src/components/ConnectWallet/index.tsx
Comment (zzmp) [inline src/components/ConnectWallet/index.tsx:15]: Can you add a comment here that this will immediately trigger lazy loading of the wallets.

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (zzmp) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx:113]: This is being recreated on each render, and then the value of the provider is being instantiated on each render.

Instead, make updateWeb3 a callback
```
const updateWeb3 = useCallback((updateContext: Web3ContextType) => {
```

and memoize the value so it is not recreated every time the provider renders
```
const value = useMemo(() => ({ web3, updateWeb3 }), [web3, updateWeb3])
return <Web3Context.Provider value={value}>{children}</Web3Context.Provider>
```

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (zzmp) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:123]: The colors on the QR code should match the theme, and not just be white/black.
I'm fine with this being done in a follow-up PR, but it should be done before merging to `main`.

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (vm) [inline src/hooks/connectWeb3/useActiveWeb3React.tsx:170]: Do we need to use this hook instead of using the `Web3ReactProvider` and `useConnector` hook that's provided by `web3-react`? I think that could reduce a lot of the complexity

Comment (vm) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx:170]: It would also remove the need for `useActiveWalletProvider`

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (vm) [inline src/hooks/connectWeb3/useActiveWeb3React.tsx:82]: if the connectors are completely memoized I would move them outside of the component, though I think we could switch to using the `Web3ReactProvider` as commented below

File: src/hooks/connectWeb3/useActiveWeb3React.tsx
Comment (kristiehuang) [reply src/hooks/connectWeb3/useActiveWeb3React.tsx:98]: this gets handled a couple lines up

```
const activeProvider = useActiveWalletProvider() // which uses the priority selector
const wallet = useMemo(() => getWallet(propsProvider ?? activeProvider), [propsProvider, activeProvider])
```

File: src/components/ConnectWallet/index.tsx
Comment (kristiehuang) [reply src/components/ConnectWallet/index.tsx:15]: @zzmp is that different from the `// Attempt to connect eagerly on mount` comment?

@vm if the user isn't already connected to the page, I don't think we want to call activate unless the user clicks `Connect wallet`

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (zzmp) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:125]: The viewbox is set to `0 0 53 53`, so the width must be a multiple of 53 (ie 106) to prevent smushing the image (sorry idk the technical word). Does `QRCode` have an option to use `50`?

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/components/ConnectWallet/ConnectWalletDialog.tsx
Comment (zzmp) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:142]: Instead of doing this dangerously, can you output an image (using QRCode.toDataUrl) and then set the src of an img element?

**Pull request merged**
